File: LeetCode/cpp/1026.cpp
class Solution {
    int diff;
public:
    int maxAncestorDiff(TreeNode *root) {
        if (!root)
            return 0;
        this->diff = 0;
        Traverse(root, root->val, root->val);
        return this->diff;
    }

    void Traverse(TreeNode *node, int min_val, int max_val) {
        if (!node)
            return;
        this->diff = max(this->diff, max(abs(node->val - min_val), abs(node->val - max_val)));
        min_val = min(min_val, node->val), max_val = max(max_val, node->val);
        Traverse(node->left, min_val, max_val);
        Traverse(node->right, min_val, max_val);
    }
};

File: LeetCode/python3/1026.py
class Solution:
    def maxAncestorDiff(self, root: TreeNode) -> int:
        if not root:
            return 0
        self.val = 0
        self.Traverse(root, root.val, root.val)
        return self.val
        
    def Traverse(self, node: TreeNode, low: int, high: int) -> None:
        if not node:
            return
        self.val = max(self.val, max(abs(node.val - low), abs(node.val - high)))
        low = min(low, node.val)
        high = max(high, node.val)
        self.Traverse(node.left, low, high)
        self.Traverse(node.right, low, high)